# RESULTS

## MODULES & LIBRARIES

In [335]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [336]:
import pandas as pd
import numpy as np
import datetime as dt
#import pandas_profiling
from math import *

import sklearn
import matplotlib
import scipy
import statsmodels
import mlxtend
import seaborn
import xgboost

In [337]:
# IMPORT MODULES

# Base -----------------------------------------------------------
import numpy as np
import pandas as pd
from scipy.stats import norm
from datetime import datetime

# Viz ------------------------------------------------------------
import matplotlib.pyplot as plt
import matplotlib.colors as ma
import matplotlib.font_manager as fm
import seaborn as sns
import plotly
import plotnine
from plotnine import *
#import patchworklib as pw

%matplotlib inline

In [338]:
import time
from datetime import timedelta

In [339]:
import matplotlib.pyplot as plt
import seaborn as sns
import plotly
%matplotlib inline

## DEF. MODELO ML (XGBoost - BID)

XGBOOST (BID MODEL - SP)
______________________________
Avg. RSME: 9.66
Avg. MAE: 6.77
Avg. REAL BID PRICE: 41.87
Avg. MAPE: 16.16 %
______________________________
Avg. RSME: 17.34
Avg. MAE: 11.77
Avg. REAL BID PRICE: 41.87
Avg. MAPE: 28.12 %
______________________________
Avg. RSME: 12.06
Avg. MAE: 8.42
Avg. REAL BID PRICE: 41.87
Avg. MAPE: 20.11 %

XGBOOST (BID MODEL - SP PAVG. PRICE MODEL)
______________________________
Avg. RSME: 21.93
Avg. MAE: 18.16
Avg. REAL BID PRICE: 45.28
Avg. MAPE: -60.08 %
______________________________
Avg. RSME: 25.11
Avg. MAE: 16.15
Avg. REAL BID PRICE: 45.28
Avg. MAPE: 23.14 %
______________________________
Avg. RSME: 12.31
Avg. MAE: 9.29
Avg. REAL BID PRICE: 45.28
Avg. MAPE: -74.00 %

### - XGBOOST (BID MODEL - SP)

In [ ]:
file_paths = ['/content/drive/MyDrive/QEnergy/Results/XGBoost/XGBoost BID MODEL OPT. - SP (scores MAPE).csv',
              '/content/drive/MyDrive/QEnergy/Results/XGBoost/Archive/Old/XGBoost BID MODEL - SP (scores)+.csv',
              '/content/drive/MyDrive/QEnergy/Results/XGBoost/Archive/Old/XGBoost BID MODEL - SP (scores)+2.csv']

In [ ]:

print('XGBOOST (BID MODEL - SP)')

for x in file_paths:
  bidsp_scores = pd.read_csv(x)

  bidsp_rsme = bidsp_scores['rmse_test'].mean()
  bidsp_mae = bidsp_scores['mae_test'].mean()
  bidsp_meanreal = bidsp_scores['meanreal_test'].mean()

  bidsp_mape = (bidsp_scores['mae_test']/bidsp_scores['meanreal_test']).mean() * 100
  #bidsp_mape = (bidsp_mae/bidsp_meanreal) * 100

  print('_'*30)
  print(f'Avg. RSME: {bidsp_rsme:0.2f}')
  print(f'Avg. MAE: {bidsp_mae:0.2f}')
  print(f'Avg. REAL BID PRICE: {bidsp_meanreal:0.2f}')
  print(f'Avg. MAPE: {bidsp_mape:0.2f} %')


XGBOOST (BID MODEL - SP)
______________________________
Avg. RSME: 9.66
Avg. MAE: 6.77
Avg. REAL BID PRICE: 41.87
Avg. MAPE: 16.14 %
______________________________
Avg. RSME: 17.34
Avg. MAE: 11.77
Avg. REAL BID PRICE: 41.87
Avg. MAPE: 28.76 %
______________________________
Avg. RSME: 12.06
Avg. MAE: 8.42
Avg. REAL BID PRICE: 41.87
Avg. MAPE: 22.45 %


In [ ]:
file_path = '/content/drive/MyDrive/QEnergy/Results/XGBoost/XGBoost BID MODEL OPT. - SP (feat.importances).csv'
bidsp_imp = pd.read_csv(file_path)

In [ ]:
bidsp_imp

feature   imp_sum  imp_2023-06-01  imp_2023-06-03  \
0       sp3_6_pavgprice  0.254414        0.258679        0.248824   
1   BMU Fuel Type_label  0.173433        0.168668        0.168831   
2        BMU Type_label  0.076637        0.062321        0.059896   
3                    Z1  0.047254        0.046077        0.049844   
4                    Z6  0.044066        0.056149        0.051983   
..                  ...       ...             ...             ...   
68                   Z9  0.000898        0.001058        0.000702   
69   sp6_12_totalvolume  0.000871        0.001407        0.000846   
70                  Z15  0.000663        0.000602        0.000566   
71       FPN Flag_label  0.000651        0.000713        0.000760   
72                  Z10  0.000595        0.000633        0.000734   

    imp_2023-06-05  imp_2023-06-07  imp_2023-06-09  imp_2023-06-11  \
0         0.242066        0.254973        0.257536        0.249316   
1         0.178096        0.182094        0.189267        0.177944   
2         0.073401        0.068626        0.066780        0.065364   
3         0.052210        0.045598        0.050131        0.048080   
4         0.047375        0.048351        0.055007        0.050557   
..             ...             ...             ...             ...   
68        0.001016        0.000758        0.001199        0.000690   
69        0.000877        0.000827        0.000834        0.000715   
70        0.000797        0.000594        0.000766        0.001574   
71        0.000831        0.000875        0.000277        0.001114   
72        0.000505        0.000355        0.000560        0.000612   

    imp_2023-06-13  imp_2023-06-15  imp_2023-06-17  imp_2023-06-19  \
0         0.255785        0.245081        0.263170        0.249544   
1         0.182774        0.179137        0.173641        0.159558   
2         0.070965        0.067646        0.078960        0.089345   
3         0.050410        0.046509        0.049732        0.043131   
4         0.050050        0.054189        0.046121        0.034000   
..             ...             ...             ...             ...   
68        0.001444        0.000929        0.000746        0.000851   
69        0.000703        0.000705        0.000746        0.000902   
70        0.000916        0.001210        0.000508        0.000419   
71        0.000323        0.000979        0.000761        0.000568   
72        0.000658        0.000682        0.000629        0.000859   

    imp_2023-06-21  imp_2023-06-23  imp_2023-06-25  imp_2023-06-27  \
0         0.252015        0.263144        0.252538        0.257788   
1         0.171999        0.172739        0.167774        0.160408   
2         0.086466        0.084821        0.098229        0.088849   
3         0.039876        0.051138        0.047729        0.040196   
4         0.032219        0.033809        0.037467        0.029863   
..             ...             ...             ...             ...   
68        0.000836        0.000819        0.000809        0.001048   
69        0.000964        0.000958        0.000894        0.000746   
70        0.000508        0.000380        0.000226        0.000586   
71        0.000379        0.000451        0.000656        0.000411   
72        0.000425        0.000607        0.000389        0.000876   

    imp_2023-06-29  
0         0.265755  
1         0.168561  
2         0.087883  
3         0.048151  
4         0.033845  
..             ...  
68        0.000565  
69        0.000936  
70        0.000294  
71        0.000670  
72        0.000400  

[73 rows x 17 columns]

In [ ]:
df = bidsp_scores['mae_test']/bidsp_scores['meanreal_test']

In [ ]:
df[df>0.9]

47     1.287901
95     1.387961
104    5.814592
106    0.957573
dtype: float64

In [ ]:
bidsp_scores.sort_values(by=['rmse_test'], ascending=False).head(40)

test_day  test_sp  size_train  rmse_train  mae_train  meanreal_train  \
53   2023-06-11       48      731255   18.570917  10.455869       30.319659   
108  2023-06-25        4      738322   18.694149  10.488042       30.608812   
82   2023-06-19       10      734764   18.901311  10.576719       30.482547   
84   2023-06-19       22      734957   18.687787  10.406250       30.488139   
48   2023-06-11       22      731069   18.557654  10.379488       30.290972   
113  2023-06-25       31      738850   18.689756  10.464675       30.553600   
47   2023-06-11       16      731009   18.647438  10.395710       30.291323   
112  2023-06-25       25      738762   19.031916  10.572229       30.558406   
122  2023-06-27       31      739993   18.661383  10.574017       30.522223   
44   2023-06-09       48      730391   18.536210  10.424312       30.320953   
107  2023-06-23       48      737764   18.607848  10.467268       30.601555   
118  2023-06-27       10      739732   18.943470  10.541190       30.519826   
52   2023-06-11       40      731203   18.489070  10.369926       30.292824   
100  2023-06-23       10      737239   18.928771  10.540225       30.591917   
5    2023-06-01       31      726410   18.689004  10.401173       30.282561   
111  2023-06-25       22      738707   18.682440  10.492369       30.564033   
126  2023-06-29        4      740861   18.913513  10.546970       30.535139   
45   2023-06-11        4      730952   18.525058  10.334987       30.291869   
26   2023-06-05       48      728621   18.717040  10.438485       30.287805   
105  2023-06-23       37      737550   18.853655  10.592399       30.603206   
134  2023-06-29       48      741620   18.812981  10.451636       30.546566   
117  2023-06-27        4      739698   18.809442  10.554216       30.520302   
86   2023-06-19       31      735042   18.935790  10.535528       30.488842   
115  2023-06-25       40      738884   18.970437  10.550140       30.553386   
54   2023-06-13        4      731943   19.204065  10.607042       30.347802   
116  2023-06-25       48      738922   18.962194  10.611301       30.553499   
12   2023-06-03       22      727522   18.682840  10.478616       30.288844   
123  2023-06-27       37      740076   18.728182  10.631674       30.522469   
88   2023-06-19       40      735226   19.303696  10.529233       30.492578   
13   2023-06-03       25      727557   18.788689  10.469185       30.288187   
71   2023-06-15       48      733305   18.582464  10.428637       30.411052   
125  2023-06-27       48      740220   18.662039  10.534214       30.522705   
38   2023-06-09       16      730113   18.424701  10.448830       30.315071   
17   2023-06-03       48      727765   18.624377  10.432554       30.287352   
103  2023-06-23       25      737411   18.747098  10.540213       30.600019   
11   2023-06-03       16      727439   18.809942  10.431573       30.289367   
95   2023-06-21       31      736431   18.738392  10.549588       30.551185   
93   2023-06-21       22      736333   18.906013  10.500835       30.548086   
110  2023-06-25       16      738585   18.720667  10.504233       30.576889   
3    2023-06-01       22      726309   18.805939  10.548393       30.281770   

     meanpred_train  resid_train  absresidmean_train  residmean_train  \
53        30.319659  -191.329608           10.455869        -0.000262   
108       30.608812   -76.821590           10.488042        -0.000104   
82        30.482547  -294.329565           10.576719        -0.000401   
84        30.488139  -183.292726           10.406250        -0.000249   
48        30.290972  -304.881688           10.379488        -0.000417   
113       30.553600  -149.447401           10.464675        -0.000202   
47        30.291323  -243.207495           10.395710        -0.000333   
112       30.558406  -214.685120           10.572229        -0.000291   
122       30.522223    35.686749           10.574017         0.000048   
44        30.320953  -369.885157     

In [ ]:
bidsp_scores.sort_values(by=['rmse_test'], ascending=False).tail(40)

test_day  test_sp  size_train  rmse_train  mae_train  meanreal_train  \
78   2023-06-17       37      734080   18.523996  10.433431       30.451645   
42   2023-06-09       37      730289   18.392027  10.354356       30.318505   
64   2023-06-15       10      732925   19.072551  10.590363       30.388155   
68   2023-06-15       31      733174   18.669314  10.397195       30.399961   
80   2023-06-17       48      734181   18.751515  10.493875       30.456926   
8    2023-06-01       48      726681   18.545819  10.471232       30.284724   
130  2023-06-29       25      741211   18.948133  10.509029       30.539505   
36   2023-06-09        4      729999   18.490069  10.402652       30.312223   
32   2023-06-07       31      729191   18.336448  10.276018       30.299456   
33   2023-06-07       37      729256   18.571411  10.392528       30.300830   
0    2023-06-01        4      726113   18.337719  10.378467       30.281333   
73   2023-06-17       10      733790   18.710647  10.441123       30.438187   
9    2023-06-03        4      727267   18.516088  10.385288       30.288282   
99   2023-06-23        4      737163   18.574508  10.402593       30.588456   
49   2023-06-11       25      731098   18.615272  10.435857       30.291405   
102  2023-06-23       22      737372   18.776888  10.506556       30.598198   
67   2023-06-15       25      733141   18.669650  10.397481       30.398146   
27   2023-06-07        4      728974   18.682945  10.463739       30.294821   
101  2023-06-23       16      737334   18.777312  10.506906       30.596392   
2    2023-06-01       16      726256   18.691094  10.551424       30.281634   
43   2023-06-09       40      730314   18.391732  10.354093       30.319155   
28   2023-06-07       10      729025   18.948559  10.574507       30.295921   
81   2023-06-19        4      734701   18.734372  10.454012       30.479652   
119  2023-06-27       16      739813   19.024174  10.566811       30.519908   
92   2023-06-21       16      736254   18.790874  10.550629       30.543837   
72   2023-06-17        4      733762   18.515047  10.426917       30.436777   
69   2023-06-15       37      733230   18.644009  10.520292       30.403146   
34   2023-06-07       40      729267   18.571275  10.392408       30.301090   
65   2023-06-15       16      733011   18.751409  10.556932       30.391645   
97   2023-06-21       40      736555   18.804010  10.507199       30.555028   
66   2023-06-15       22      733110   18.670067  10.397866       30.396598   
39   2023-06-09       22      730180   18.540967  10.480686       30.316794   
37   2023-06-09       10      730043   18.620186  10.454745       30.313272   
63   2023-06-15        4      732905   18.880645  10.522243       30.387229   
60   2023-06-13       37      732376   18.567632  10.553278       30.365134   
90   2023-06-21        4      736101   18.774438  10.546954       30.535531   
31   2023-06-07       25      729138   18.818266  10.466948       30.298478   
30   2023-06-07       22      729138   18.818266  10.466948       30.298478   
46   2023-06-11       10      730964   18.405500  10.355410       30.291424   
114  2023-06-25       37      738862   19.023366  10.643598       30.553248   

     meanpred_train  resid_train  absresidmean_train  residmean_train  \
78        30.451645   401.778275           10.433431         0.000547   
42        30.318505  -644.833080           10.354356        -0.000883   
64        30.388155  -127.363733           10.590363        -0.000174   
68        30.399961   181.314956           10.397195         0.000247   
80        30.456926  -268.437737           10.493875        -0.000366   
8         30.284724  -252.957190           10.471232        -0.000348   
130       30.539505    32.008672           10.509029         0.000043   
36        30.312223    48.017846           10.402652         0.000066   
32        30.299456  -493.903830           10.276018        -0.000677   
33        30.300830  -102.790188     

In [ ]:
bidsp_scores.pivot_table(index=['test_day'], aggfunc={'rmse_test':['mean','max','min'], 'size_test':'sum'})

rmse_test                       size_test
                  max       mean        min       sum
test_day                                             
2023-06-01  21.029583  10.256216   3.853872       111
2023-06-03  17.735827  12.794569   4.382089       105
2023-06-05  19.848419  11.829142   6.388659        99
2023-06-07  10.619110   5.205678   2.046849        63
2023-06-09  24.707636   9.474342   2.328091        71
2023-06-11  51.826469  21.946885   4.030455        70
2023-06-13  18.276598   9.348959   2.057191       133
2023-06-15  17.453984   6.006169   2.213072        65
2023-06-17  12.027221   7.863615   3.585177        66
2023-06-19  34.660594  16.517258   3.783363       145
2023-06-21  16.241944   8.607557   2.047326        93
2023-06-23  22.791374  13.317884   3.939228       128
2023-06-25  37.670097  22.855305  14.974794       126
2023-06-27  25.871569  15.125698   3.736691       108
2023-06-29  20.503555  11.270649   4.972278       172

In [ ]:
bidsp_scores.pivot_table(index=['test_sp'], aggfunc={'rmse_test':['mean','max','min'], 'size_test':'sum'})

rmse_test                      size_test
               max       mean       min       sum
test_sp                                          
4        37.670097  10.423584  2.047326       142
10       34.660594  12.140027  2.328091       156
16       29.328041  10.128056  3.138409       213
22       31.580508  13.534181  2.354921       170
25       27.403314  11.121414  2.046849       147
31       29.864251  13.943913  4.726080       165
37       19.312996   9.242049  2.057191       188
40       22.271032  10.806102  2.880809       197
48       51.826469  17.106523  5.571116       177

### - XGBOOST (BID MODEL - SP PAVG. PRICE MODEL)

In [ ]:
file_paths = ['/content/drive/MyDrive/QEnergy/Results/XGBoost/XGBoost BID SP PAVG. PRICE MODEL OPT. - SP (scores MAPE).csv',
              '/content/drive/MyDrive/QEnergy/Results/XGBoost/Archive/Old/XGBoost BID SP PAVG. PRICE MODEL - SP (scores)+.csv',
              '/content/drive/MyDrive/QEnergy/Results/XGBoost/Archive/Old/XGBoost BID SP PAVG. PRICE MODEL - SP (scores)+2.csv']

In [ ]:

print('XGBOOST (BID MODEL - SP PAVG. PRICE MODEL)')

for x in file_paths:

  bidpavg_scores = pd.read_csv(x)

  bidpavg_rsme = bidpavg_scores['rmse_test'].mean()
  bidpavg_mae = bidpavg_scores['mae_test'].mean()
  bidpavg_meanreal = bidpavg_scores['meanreal_test'].mean()

  bidpavg_mape = (bidpavg_scores['mae_test']/bidpavg_scores['meanreal_test']).mean() * 100
  #bidpavg_mape = (bidpavg_mae/bidpavg_meanreal) * 100

  print('_'*30)
  print(f'Avg. RSME: {bidpavg_rsme:0.2f}')
  print(f'Avg. MAE: {bidpavg_mae:0.2f}')
  print(f'Avg. REAL BID PRICE: {bidpavg_meanreal:0.2f}')
  print(f'Avg. MAPE: {bidpavg_mape:0.2f} %')


XGBOOST (BID MODEL - SP PAVG. PRICE MODEL)
______________________________
Avg. RSME: 21.93
Avg. MAE: 18.16
Avg. REAL BID PRICE: 45.28
Avg. MAPE: -60.08 %
______________________________
Avg. RSME: 25.11
Avg. MAE: 16.15
Avg. REAL BID PRICE: 45.28
Avg. MAPE: 23.14 %
______________________________
Avg. RSME: 12.31
Avg. MAE: 9.29
Avg. REAL BID PRICE: 45.28
Avg. MAPE: -74.00 %


In [ ]:
bidpavg_scores = pd.read_csv(file_paths[0])
bidpavg_scores.sort_values(by=['mae_test'], ascending=False)

test_day  test_sp  size_train  rmse_train  mae_train  meanreal_train  \
110  2023-06-25       16      235938   74.601396  26.844671       72.640875   
109  2023-06-25       10      235904   74.598060  26.835283       72.656736   
112  2023-06-25       25      235964   74.600967  26.847080       72.631260   
53   2023-06-11       48      232269   75.138198  27.160018       72.825193   
45   2023-06-11        4      232090   73.437092  26.960681       72.788187   
..          ...      ...         ...         ...        ...             ...   
49   2023-06-11       25      232165   73.430088  26.961332       72.775418   
43   2023-06-09       40      231834   73.551781  27.089225       72.847717   
30   2023-06-07       22      231227   73.649965  27.081509       72.919971   
46   2023-06-11       10      232097   73.436596  26.961133       72.786736   
114  2023-06-25       37      236000   74.590422  26.845682       72.627526   

     meanpred_train  resid_train  absresidmean_train  residmean_train  \
110       72.640875     1.519427           26.844671         0.000006   
109       72.656736     1.552897           26.835283         0.000007   
112       72.631260     1.463016           26.847080         0.000006   
53        72.825193    12.085751           27.160018         0.000052   
45        72.788187   141.122352           26.960681         0.000608   
..              ...          ...                 ...              ...   
49        72.775418   141.091976           26.961332         0.000608   
43        72.847717    53.382165           27.089225         0.000230   
30        72.919971     4.636825           27.081509         0.000020   
46        72.786736   141.228303           26.961133         0.000608   
114       72.627526    -1.081143           26.845682        -0.000005   

     stdresid_train  size_test   rmse_test   mae_test  meanreal_test  \
110        0.000316          3  103.012031  82.199234    -105.365296   
109        0.000316          4  104.317341  82.130300     -85.250531   
112        0.000316          3   79.675066  68.804263      36.732410   
53         0.000323          4   67.194501  64.485618      53.493750   
45         0.000316          1   62.702736  62.702736      21.014805   
..              ...        ...         ...        ...            ...   
49         0.000316          2    3.527012   3.162125      30.409693   
43         0.000317          4    2.658696   2.283674      50.328665   
30         0.000319          0         NaN        NaN            NaN   
46         0.000316          0         NaN        NaN            NaN   
114        0.000316          0         NaN        NaN            NaN   

     meanpred_test  resid_test  absresidmean_test  residmean_test  \
110     -23.166061 -246.597701          82.199234      -82.199234   
109     -22.145018 -252.422057          82.130300      -63.105514   
112     -32.071854  206.412789          68.804263       68.804263   
53      117.979370 -257.942473          64.485618      -64.485618   
45      -41.687931   62.702736          62.702736       62.702736   
..             ...         ...                ...             ...   
49       27.247568    6.324250           3.162125        3.162125   
43       49.615112    2.854216           2.283674        0.713554   
30             NaN         NaN                NaN             NaN   
46             NaN         NaN                NaN             NaN   
114            NaN         NaN                NaN             NaN   

     stdresid_test  
110      20.695583  
109      20.766261  
112      13.391911  
53        4.721666  
45        0.000000  
..             ...  
49        0.781150  
43        0.640288  
30             NaN  
46             NaN  
114            NaN  

[135 rows x 20 columns]

In [ ]:
bidpavg_scores

test_day  test_sp  size_train  rmse_train  mae_train  meanreal_train  \
0    2023-06-01        4      229670   73.868352  27.140193       73.173249   
1    2023-06-01       10      229700   73.863878  27.138697       73.167785   
2    2023-06-01       16      229743   73.858186  27.137925       73.160284   
3    2023-06-01       22      229782   73.852707  27.136366       73.153514   
4    2023-06-01       25      229808   73.849109  27.135165       73.149549   
..          ...      ...         ...         ...        ...             ...   
130  2023-06-29       25      237099   74.490246  26.888219       72.447233   
131  2023-06-29       31      237161   74.481078  26.884089       72.442964   
132  2023-06-29       37      237216   74.473113  26.881636       72.439203   
133  2023-06-29       40      237231   74.470948  26.881079       72.437787   
134  2023-06-29       48      237289   74.462773  26.878017       72.432964   

     meanpred_train  resid_train  absresidmean_train  residmean_train  \
0         73.173249     0.893549           27.140193         0.000004   
1         73.167785     0.634567           27.138697         0.000003   
2         73.160284     1.140961           27.137925         0.000005   
3         73.153514     1.183928           27.136366         0.000005   
4         73.149549     1.463546           27.135165         0.000006   
..              ...          ...                 ...              ...   
130       72.447233     1.679740           26.888219         0.000007   
131       72.442964     1.596234           26.884089         0.000007   
132       72.439203     1.810977           26.881636         0.000008   
133       72.437787     1.285373           26.881079         0.000005   
134       72.432964     1.736711           26.878017         0.000007   

     stdresid_train  size_test  rmse_test   mae_test  meanreal_test  \
0          0.000322          5   9.011282   7.935806      30.074272   
1          0.000322          6  23.619800  22.809780      32.191245   
2          0.000321          6  14.349966  13.065280      30.040695   
3          0.000321          7  10.733756   8.481880      29.932998   
4          0.000321          7  14.749487  10.377737      29.423439   
..              ...        ...        ...        ...            ...   
130        0.000314          8  11.580501   7.421622      53.973159   
131        0.000314          9  27.532739  21.293028      57.624310   
132        0.000314          7  15.941256  14.915366      53.403294   
133        0.000314          6  16.236559  13.955520      52.449252   
134        0.000314         10  25.718279  21.040678      40.707818   

     meanpred_test  resid_test  absresidmean_test  residmean_test  \
0        22.138468   39.679029           7.935806        7.935806   
1         9.381466  136.858678          22.809780       22.809780   
2        17.374048   75.999884          13.065280       12.666647   
3        28.390854   10.795013           8.481880        1.542145   
4        23.621286   40.615077          10.377737        5.802154   
..             ...         ...                ...             ...   
130      54.912834   -7.517405           7.421622       -0.939676   
131      41.658634  143.691081          21.293028       15.965676   
132      44.327709   63.529085          14.915366        9.075584   
133      47.960144   26.934656          13.955520        4.489109   
134      38.385353   23.224640          21.040678        2.322464   

     stdresid_test  
0         0.853843  
1         1.022102  
2         1.123951  
3         1.517485  
4         1.937189  
..             ...  
130       1.442789  
131       2.492331  
132       1.872232  
133       2.600608  
134       2.561320  

[135 rows x 20 columns]

In [ ]:
file_path = ''
file_path = '/content/drive/MyDrive/QEnergy/Results/XGBoost/XGBoost BID SP PAVG. PRICE MODEL OPT. - SP (feat.importances).csv'
bidpavg_imp = pd.read_csv(file_path)

In [ ]:
bidpavg_imp

feature   imp_sum  imp_2023-06-01  imp_2023-06-03  \
0                 sp3_6_pavgprice  0.158691        0.158440        0.158981   
1                      Coal_price  0.134116        0.134891        0.135189   
2                        DAMargin  0.121318        0.120908        0.122211   
3                sp6_12_pavgprice  0.098127        0.097974        0.098405   
4                       Month_SIN  0.079448        0.079297        0.079626   
5                  BrentOil_price  0.055219        0.055759        0.055707   
6         WindOffshore_Generation  0.039805        0.040367        0.040290   
7          WindOnshore_Generation  0.034143        0.032135        0.032250   
8                     DATF_Demand  0.033818        0.033246        0.033621   
9                     DAImbalance  0.028200        0.028265        0.028447   
10                LOC Center LONG  0.028168        0.026610        0.026606   
11       BMU GSP Group Name_label  0.027444        0.027848        0.027849   
12                    Month_ordin  0.022367        0.022222        0.022386   
13              sp3_6_totalvolume  0.022322        0.023996        0.024070   
14               marketIndexPrice  0.021855        0.021491        0.021648   
15             sp6_12_totalvolume  0.019369        0.019303        0.019376   
16               sp3_6_meanvolume  0.011682        0.014716        0.014566   
17                    DANF_Demand  0.009694        0.011295        0.009156   
18                 settPeriod_SIN  0.007523        0.006892        0.004998   
19    SystemTotalLoad_Consumption  0.006880        0.006802        0.006813   
20           settlementDate_ordin  0.005476        0.004781        0.004873   
21                 DAID_spnDemand  0.005257        0.004534        0.004550   
22                 LOC Center LAT  0.005088        0.005097        0.005103   
23             DAIG_spnGeneration  0.004356        0.004151        0.004017   
24         settlementPeriod_ordin  0.003685        0.003608        0.003705   
25  GenerationForecast_Production  0.002626        0.002722        0.002662   
26                    Weekday_SIN  0.002269        0.002261        0.002344   
27              sp6_12_meanvolume  0.002088        0.002084        0.002071   
28                        Day_SIN  0.002039        0.002002        0.001986   
29                 settPeriod_COS  0.001779        0.001574        0.001470   
30                        Holiday  0.001464        0.001885        0.001953   
31                        Day_COS  0.001319        0.001256        0.001571   
32              marketIndexVolume  0.000994        0.000277        0.000257   
33                      Day_ordin  0.000771        0.000791        0.000803   
34               Solar_Generation  0.000602        0.000523        0.000441   
35                    Weekday_COS  0.000000        0.000000        0.000000   
36                      Month_COS  0.000000        0.000000        0.000000   
37                     activeFlag  0.000000        0.000000        0.000000   
38                   NatGas_price  0.000000        0.000000        0.000000   
39                        Weekend  0.000000        0.000000        0.000000   

    imp_2023-06-05  imp_2023-06-07  imp_2023-06-09  imp_2023-06-11  \
0         0.159032        0.158562        0.159183        0.157737   
1         0.134904        0.134417        0.135287        0.133944   
2         0.122294        0.121787        0.122322        0.122491   
3         0.098379        0.097965        0.098516        0.097729   
4         0.079688        0.079506        0.079950        0.079049   
5         0.055519        0.055206        0.055430        0.054895   
6         0.039998        0.039751        0.040003        0.039793   
7         0.032163        0.032135        0.032215        0.039419   
8         0.033778        0.033796        0.034051        0.033790   
9         0.028354        0.028230        0.028274        0.028060   
10        0.0264

In [ ]:
bidsp_scores.sort_values(by=['rmse_test'], ascending=False).head(40)

test_day  test_sp  size_train  rmse_train  mae_train  meanreal_train  \
53   2023-06-11       48      731255   18.570917  10.455869       30.319659   
108  2023-06-25        4      738322   18.694149  10.488042       30.608812   
82   2023-06-19       10      734764   18.901311  10.576719       30.482547   
84   2023-06-19       22      734957   18.687787  10.406250       30.488139   
48   2023-06-11       22      731069   18.557654  10.379488       30.290972   
113  2023-06-25       31      738850   18.689756  10.464675       30.553600   
47   2023-06-11       16      731009   18.647438  10.395710       30.291323   
112  2023-06-25       25      738762   19.031916  10.572229       30.558406   
122  2023-06-27       31      739993   18.661383  10.574017       30.522223   
44   2023-06-09       48      730391   18.536210  10.424312       30.320953   
107  2023-06-23       48      737764   18.607848  10.467268       30.601555   
118  2023-06-27       10      739732   18.943470  10.541190       30.519826   
52   2023-06-11       40      731203   18.489070  10.369926       30.292824   
100  2023-06-23       10      737239   18.928771  10.540225       30.591917   
5    2023-06-01       31      726410   18.689004  10.401173       30.282561   
111  2023-06-25       22      738707   18.682440  10.492369       30.564033   
126  2023-06-29        4      740861   18.913513  10.546970       30.535139   
45   2023-06-11        4      730952   18.525058  10.334987       30.291869   
26   2023-06-05       48      728621   18.717040  10.438485       30.287805   
105  2023-06-23       37      737550   18.853655  10.592399       30.603206   
134  2023-06-29       48      741620   18.812981  10.451636       30.546566   
117  2023-06-27        4      739698   18.809442  10.554216       30.520302   
86   2023-06-19       31      735042   18.935790  10.535528       30.488842   
115  2023-06-25       40      738884   18.970437  10.550140       30.553386   
54   2023-06-13        4      731943   19.204065  10.607042       30.347802   
116  2023-06-25       48      738922   18.962194  10.611301       30.553499   
12   2023-06-03       22      727522   18.682840  10.478616       30.288844   
123  2023-06-27       37      740076   18.728182  10.631674       30.522469   
88   2023-06-19       40      735226   19.303696  10.529233       30.492578   
13   2023-06-03       25      727557   18.788689  10.469185       30.288187   
71   2023-06-15       48      733305   18.582464  10.428637       30.411052   
125  2023-06-27       48      740220   18.662039  10.534214       30.522705   
38   2023-06-09       16      730113   18.424701  10.448830       30.315071   
17   2023-06-03       48      727765   18.624377  10.432554       30.287352   
103  2023-06-23       25      737411   18.747098  10.540213       30.600019   
11   2023-06-03       16      727439   18.809942  10.431573       30.289367   
95   2023-06-21       31      736431   18.738392  10.549588       30.551185   
93   2023-06-21       22      736333   18.906013  10.500835       30.548086   
110  2023-06-25       16      738585   18.720667  10.504233       30.576889   
3    2023-06-01       22      726309   18.805939  10.548393       30.281770   

     meanpred_train  resid_train  absresidmean_train  residmean_train  \
53        30.319659  -191.329608           10.455869        -0.000262   
108       30.608812   -76.821590           10.488042        -0.000104   
82        30.482547  -294.329565           10.576719        -0.000401   
84        30.488139  -183.292726           10.406250        -0.000249   
48        30.290972  -304.881688           10.379488        -0.000417   
113       30.553600  -149.447401           10.464675        -0.000202   
47        30.291323  -243.207495           10.395710        -0.000333   
112       30.558406  -214.685120           10.572229        -0.000291   
122       30.522223    35.686749           10.574017         0.000048   
44        30.320953  -369.885157     

In [ ]:
bidsp_scores.sort_values(by=['rmse_test'], ascending=False).tail(40)

test_day  test_sp  size_train  rmse_train  mae_train  meanreal_train  \
78   2023-06-17       37      734080   18.523996  10.433431       30.451645   
42   2023-06-09       37      730289   18.392027  10.354356       30.318505   
64   2023-06-15       10      732925   19.072551  10.590363       30.388155   
68   2023-06-15       31      733174   18.669314  10.397195       30.399961   
80   2023-06-17       48      734181   18.751515  10.493875       30.456926   
8    2023-06-01       48      726681   18.545819  10.471232       30.284724   
130  2023-06-29       25      741211   18.948133  10.509029       30.539505   
36   2023-06-09        4      729999   18.490069  10.402652       30.312223   
32   2023-06-07       31      729191   18.336448  10.276018       30.299456   
33   2023-06-07       37      729256   18.571411  10.392528       30.300830   
0    2023-06-01        4      726113   18.337719  10.378467       30.281333   
73   2023-06-17       10      733790   18.710647  10.441123       30.438187   
9    2023-06-03        4      727267   18.516088  10.385288       30.288282   
99   2023-06-23        4      737163   18.574508  10.402593       30.588456   
49   2023-06-11       25      731098   18.615272  10.435857       30.291405   
102  2023-06-23       22      737372   18.776888  10.506556       30.598198   
67   2023-06-15       25      733141   18.669650  10.397481       30.398146   
27   2023-06-07        4      728974   18.682945  10.463739       30.294821   
101  2023-06-23       16      737334   18.777312  10.506906       30.596392   
2    2023-06-01       16      726256   18.691094  10.551424       30.281634   
43   2023-06-09       40      730314   18.391732  10.354093       30.319155   
28   2023-06-07       10      729025   18.948559  10.574507       30.295921   
81   2023-06-19        4      734701   18.734372  10.454012       30.479652   
119  2023-06-27       16      739813   19.024174  10.566811       30.519908   
92   2023-06-21       16      736254   18.790874  10.550629       30.543837   
72   2023-06-17        4      733762   18.515047  10.426917       30.436777   
69   2023-06-15       37      733230   18.644009  10.520292       30.403146   
34   2023-06-07       40      729267   18.571275  10.392408       30.301090   
65   2023-06-15       16      733011   18.751409  10.556932       30.391645   
97   2023-06-21       40      736555   18.804010  10.507199       30.555028   
66   2023-06-15       22      733110   18.670067  10.397866       30.396598   
39   2023-06-09       22      730180   18.540967  10.480686       30.316794   
37   2023-06-09       10      730043   18.620186  10.454745       30.313272   
63   2023-06-15        4      732905   18.880645  10.522243       30.387229   
60   2023-06-13       37      732376   18.567632  10.553278       30.365134   
90   2023-06-21        4      736101   18.774438  10.546954       30.535531   
31   2023-06-07       25      729138   18.818266  10.466948       30.298478   
30   2023-06-07       22      729138   18.818266  10.466948       30.298478   
46   2023-06-11       10      730964   18.405500  10.355410       30.291424   
114  2023-06-25       37      738862   19.023366  10.643598       30.553248   

     meanpred_train  resid_train  absresidmean_train  residmean_train  \
78        30.451645   401.778275           10.433431         0.000547   
42        30.318505  -644.833080           10.354356        -0.000883   
64        30.388155  -127.363733           10.590363        -0.000174   
68        30.399961   181.314956           10.397195         0.000247   
80        30.456926  -268.437737           10.493875        -0.000366   
8         30.284724  -252.957190           10.471232        -0.000348   
130       30.539505    32.008672           10.509029         0.000043   
36        30.312223    48.017846           10.402652         0.000066   
32        30.299456  -493.903830           10.276018        -0.000677   
33        30.300830  -102.790188     

In [ ]:
bidsp_scores.pivot_table(index=['test_day'], aggfunc={'rmse_test':['mean','max','min'], 'size_test':'sum'})

rmse_test                       size_test
                  max       mean        min       sum
test_day                                             
2023-06-01  21.029583  10.256216   3.853872       111
2023-06-03  17.735827  12.794569   4.382089       105
2023-06-05  19.848419  11.829142   6.388659        99
2023-06-07  10.619110   5.205678   2.046849        63
2023-06-09  24.707636   9.474342   2.328091        71
2023-06-11  51.826469  21.946885   4.030455        70
2023-06-13  18.276598   9.348959   2.057191       133
2023-06-15  17.453984   6.006169   2.213072        65
2023-06-17  12.027221   7.863615   3.585177        66
2023-06-19  34.660594  16.517258   3.783363       145
2023-06-21  16.241944   8.607557   2.047326        93
2023-06-23  22.791374  13.317884   3.939228       128
2023-06-25  37.670097  22.855305  14.974794       126
2023-06-27  25.871569  15.125698   3.736691       108
2023-06-29  20.503555  11.270649   4.972278       172

In [ ]:
bidsp_scores.pivot_table(index=['test_sp'], aggfunc={'rmse_test':['mean','max','min'], 'size_test':'sum'})

rmse_test                      size_test
               max       mean       min       sum
test_sp                                          
4        37.670097  10.423584  2.047326       142
10       34.660594  12.140027  2.328091       156
16       29.328041  10.128056  3.138409       213
22       31.580508  13.534181  2.354921       170
25       27.403314  11.121414  2.046849       147
31       29.864251  13.943913  4.726080       165
37       19.312996   9.242049  2.057191       188
40       22.271032  10.806102  2.880809       197
48       51.826469  17.106523  5.571116       177

## DEF. MODELO ML (XGBoost - OFFER)

XGBOOST (OFFER MODEL - SP)
______________________________
Avg. RSME: 19.21
Avg. MAE: 13.14
Avg. REAL BID PRICE: 128.89
Avg. MAPE: 10.42 %
______________________________
Avg. RSME: 13.81
Avg. MAE: 11.26
Avg. REAL BID PRICE: 128.89
Avg. MAPE: 8.81 %
______________________________
Avg. RSME: 14.36
Avg. MAE: 10.57
Avg. REAL BID PRICE: 128.89
Avg. MAPE: 8.33 %
______________________________
Avg. RSME: 17.88
Avg. MAE: 12.67
Avg. REAL BID PRICE: 128.89
Avg. MAPE: 9.91 %

XGBOOST (OFFER MODEL - SP PAVG. PRICE MODEL)
______________________________
Avg. RSME: 14.80
Avg. MAE: 11.89
Avg. REAL BID PRICE: 128.45
Avg. MAPE: 9.37 %
______________________________
Avg. RSME: 12.90
Avg. MAE: 10.63
Avg. REAL BID PRICE: 128.45
Avg. MAPE: 8.31 %
______________________________
Avg. RSME: 16.27
Avg. MAE: 12.94
Avg. REAL BID PRICE: 128.45
Avg. MAPE: 10.48 %
______________________________
Avg. RSME: 13.61
Avg. MAE: 10.82
Avg. REAL BID PRICE: 128.45
Avg. MAPE: 8.60 %

### - XGBOOST (OFFER MODEL - SP)

In [352]:
file_paths = ['/content/drive/MyDrive/QEnergy/Results/XGBoost/XGBoost OFFER MODEL OPT. - SP (scores MAPE).csv',
             '/content/drive/MyDrive/QEnergy/Results/XGBoost/Archive/Old/XGBoost OFFER MODEL - SP (scores)+.csv',
             '/content/drive/MyDrive/QEnergy/Results/XGBoost/Archive/Old/XGBoost OFFER MODEL OPT. - SP (scores)+.csv',
             '/content/drive/MyDrive/QEnergy/Results/XGBoost/Archive/Old/XGBoost OFFER MODEL - SP (scores)+2.csv']

In [353]:

print('XGBOOST (OFFER MODEL - SP)')

for x in file_paths:
  offersp_scores = pd.read_csv(x)

  offersp_rsme = offersp_scores['rmse_test'].mean()
  offersp_mae = offersp_scores['mae_test'].mean()
  offersp_meanreal = offersp_scores['meanreal_test'].mean()

  offersp_mape = (offersp_scores['mae_test']/offersp_scores['meanreal_test']).mean() * 100
  #offersp_mape = (offersp_mae/offersp_meanreal) * 100

  print('_'*30)
  print(f'Avg. RSME: {offersp_rsme:0.2f}')
  print(f'Avg. MAE: {offersp_mae:0.2f}')
  print(f'Avg. REAL BID PRICE: {offersp_meanreal:0.2f}')
  print(f'Avg. MAPE: {offersp_mape:0.2f} %')


XGBOOST (OFFER MODEL - SP)
______________________________
Avg. RSME: 19.21
Avg. MAE: 13.14
Avg. REAL BID PRICE: 128.89
Avg. MAPE: 10.42 %
______________________________
Avg. RSME: 13.81
Avg. MAE: 11.26
Avg. REAL BID PRICE: 128.89
Avg. MAPE: 8.81 %
______________________________
Avg. RSME: 14.36
Avg. MAE: 10.57
Avg. REAL BID PRICE: 128.89
Avg. MAPE: 8.33 %
______________________________
Avg. RSME: 17.88
Avg. MAE: 12.67
Avg. REAL BID PRICE: 128.89
Avg. MAPE: 9.91 %


In [354]:
file_path = '/content/drive/MyDrive/QEnergy/Results/XGBoost/XGBoost OFFER MODEL OPT. - SP (feat.importances).csv'
offersp_imp = pd.read_csv(file_path)

In [355]:
offersp_imp

feature   imp_sum  imp_2023-06-01  imp_2023-06-03  imp_2023-06-05  \
0   PC Status_label  0.254341        0.400253        0.321052        0.392893   
1                Z6  0.181704        0.127367        0.145010        0.145236   
2   sp3_6_pavgprice  0.065847        0.053186        0.061810        0.053515   
3           LOC LAT  0.048346        0.041740        0.045005        0.035579   
4        Coal_price  0.038307        0.035866        0.037816        0.030798   
..              ...       ...             ...             ...             ...   
68          Weekend  0.001003        0.001955        0.000464        0.004540   
69               Z4  0.000891        0.000419        0.000767        0.000888   
70               Z2  0.000578        0.000109        0.000316        0.002360   
71          Holiday  0.000253        0.000060        0.000317        0.000160   
72   FPN Flag_label  0.000080        0.000046        0.000114        0.000058   

    imp_2023-06-07  imp_2023-06-09  imp_2023-06-11  imp_2023-06-13  \
0         0.377965        0.388497        0.403782        0.349089   
1         0.162713        0.129623        0.144736        0.179180   
2         0.058453        0.057456        0.053837        0.061041   
3         0.033785        0.035602        0.033927        0.045404   
4         0.032609        0.030854        0.038353        0.033560   
..             ...             ...             ...             ...   
68        0.000272        0.000174        0.000123        0.000283   
69        0.000911        0.001370        0.000431        0.000563   
70        0.000078        0.000306        0.000166        0.000226   
71        0.000219        0.000116        0.000301        0.000252   
72        0.000104        0.000044        0.000034        0.000003   

    imp_2023-06-15  imp_2023-06-17  imp_2023-06-19  imp_2023-06-21  \
0         0.354978        0.364637        0.453994        0.003291   
1         0.154651        0.160268        0.102511        0.182624   
2         0.067749        0.065849        0.058107        0.088462   
3         0.051782        0.052681        0.051077        0.075936   
4         0.033528        0.035779        0.033900        0.054163   
..             ...             ...             ...             ...   
68        0.000512        0.000359        0.000843        0.000513   
69        0.000499        0.000512        0.001173        0.001176   
70        0.000295        0.000345        0.000198        0.000254   
71        0.000242        0.000226        0.000217        0.000352   
72        0.000058        0.000055        0.000055        0.000101   

    imp_2023-06-23  imp_2023-06-25  imp_2023-06-27  imp_2023-06-29  
0         0.000568        0.000625        0.001490        0.001997  
1         0.272455        0.251055        0.305682        0.262449  
2         0.076356        0.075307        0.076008        0.080568  
3         0.055897        0.053286        0.055109        0.058384  
4         0.042236        0.043296        0.043129        0.048724  
..             ...             ...             ...             ...  
68        0.002118        0.000076        0.002555        0.000264  
69        0.001965        0.000561        0.001448        0.000682  
70        0.003022        0.000246        0.000436        0.000316  
71        0.000231        0.000358        0.000248        0.000500  
72        0.000092        0.000133        0.000202        0.000097  

[73 rows x 17 columns]

In [356]:
offersp_scores.sort_values(by=['rmse_test'], ascending=False).head(40)

test_day  test_sp  size_train  rmse_train  mae_train  meanreal_train  \
58   2023-06-13       25      435623     46.4390    19.9801        215.3394   
110  2023-06-25       16      440418     47.7973    19.9840        214.4896   
45   2023-06-11        4      433534     45.9261    19.8673        215.7109   
130  2023-06-29       25      443582     48.7125    20.1759        213.9454   
8    2023-06-01       48      429281     45.9836    20.0172        216.7006   
99   2023-06-23        4      439388     46.9645    19.9488        214.6854   
32   2023-06-07       31      432159     46.9621    20.2237        216.0235   
26   2023-06-05       48      431491     47.0911    20.2239        216.1877   
9    2023-06-03        4      429723     47.8341    20.2387        216.6028   
107  2023-06-23       48      439665     47.8077    19.7961        214.6326   
44   2023-06-09       48      432824     47.6662    20.0994        215.8660   
54   2023-06-13        4      435455     46.3922    19.9871        215.3705   
134  2023-06-29       48      443774     47.4762    19.9432        213.9116   
75   2023-06-17       22      437319     46.9812    19.9626        215.0573   
116  2023-06-25       48      441997     48.6666    19.9327        214.2172   
118  2023-06-27       10      442490     49.4645    20.0494        214.1282   
17   2023-06-03       48      430119     48.7694    20.3163        216.5068   
55   2023-06-13       10      435511     47.5351    20.0767        215.3604   
61   2023-06-13       40      435776     47.3760    20.1173        215.3097   
57   2023-06-13       22      435615     46.6741    20.0986        215.3414   
77   2023-06-17       31      437338     46.3861    19.9034        215.0529   
131  2023-06-29       31      443614     48.1112    20.0312        213.9392   
41   2023-06-09       31      432712     47.8251    20.1831        215.8915   
56   2023-06-13       16      435573     46.4996    19.8784        215.3500   
95   2023-06-21       31      438711     47.0648    19.9872        214.8006   
0    2023-06-01        4      428958     47.4145    20.1129        216.7766   
84   2023-06-19       22      437693     46.2132    19.8882        214.9824   
92   2023-06-21       16      438684     46.8473    19.9650        214.8055   
47   2023-06-11       16      433887     46.9030    20.1281        215.6397   
103  2023-06-23       25      439456     46.9612    19.9468        214.6722   
112  2023-06-25       25      441057     47.5652    20.0540        214.3749   
13   2023-06-03       25      429881     47.9714    20.2205        216.5656   
127  2023-06-29       10      443518     46.9425    19.8681        213.9562   
11   2023-06-03       16      429779     48.5943    20.1877        216.5910   
125  2023-06-27       48      442817     46.6780    19.8689        214.0719   
89   2023-06-19       48      437838     47.1087    19.9113        214.9549   
104  2023-06-23       31      439497     46.9592    19.9457        214.6635   
14   2023-06-03       31      429967     46.8508    20.0780        216.5437   
115  2023-06-25       40      441794     46.1543    19.7428        214.2503   
48   2023-06-11       22      433980     46.6293    20.0767        215.6211   

     meanpred_train  resid_train  absresidmean_train  residmean_train  \
58         215.3394    3156.9705             19.9801           0.0072   
110        214.4896    2718.4626             19.9840           0.0062   
45         215.7109    3119.3983             19.8673           0.0072   
130        213.9454    2088.9540             20.1759           0.0047   
8          216.7006    4375.7865             20.0172           0.0102   
99         214.6854    2560.2220             19.9488           0.0058   
32         216.0235    1377.7169             20.2237           0.0032   
26         216.1877    2085.0563             20.2239           0.0048   
9          216.6028    2478.9086             20.2387           0.0058   
107        214.6326    1799.4164     

In [357]:
offersp_scores.sort_values(by=['rmse_test'], ascending=False).tail(40)

test_day  test_sp  size_train  rmse_train  mae_train  meanreal_train  \
85   2023-06-19       25      437705     46.4952    19.9011        214.9800   
49   2023-06-11       25      434087     49.3829    20.2594        215.5993   
76   2023-06-17       25      437325     46.3868    19.9037        215.0559   
39   2023-06-09       22      432653     47.4811    20.2137        215.9079   
43   2023-06-09       40      432778     47.0865    20.1649        215.8760   
50   2023-06-11       31      434224     48.2162    20.1810        215.5714   
114  2023-06-25       37      441778     46.1550    19.7431        214.2532   
16   2023-06-03       40      430061     47.8773    20.2703        216.5204   
96   2023-06-21       37      438815     47.1604    19.8626        214.7833   
42   2023-06-09       37      432754     47.9656    20.2419        215.8817   
7    2023-06-01       40      429245     48.1989    20.3294        216.7095   
5    2023-06-01       31      429113     47.8064    20.3098        216.7409   
3    2023-06-01       22      429058     47.8428    20.3490        216.7547   
98   2023-06-21       48      439023     45.6367    19.8565        214.7505   
18   2023-06-05        4      431171     47.3744    20.2330        216.2604   
80   2023-06-17       48      437416     45.9581    19.8026        215.0386   
88   2023-06-19       40      437779     46.1907    19.8909        214.9655   
119  2023-06-27       16      442548     49.6770    20.0937        214.1196   
124  2023-06-27       40      442638     48.2226    19.9485        214.1025   
1    2023-06-01       10      428983     47.1653    20.1496        216.7719   
120  2023-06-27       22      442585     49.6750    20.0926        214.1127   
22   2023-06-05       25      431307     47.2286    20.1168        216.2289   
133  2023-06-29       40      443670     47.5099    19.9806        213.9290   
101  2023-06-23       16      439437     46.9622    19.9475        214.6757   
31   2023-06-07       25      432152     47.0878    20.1153        216.0254   
34   2023-06-07       40      432197     47.7549    20.1467        216.0142   
68   2023-06-15       31      436521     46.5057    19.8458        215.1780   
117  2023-06-27        4      442456     46.8425    19.9479        214.1342   
35   2023-06-07       48      432231     47.5137    20.1337        216.0058   
87   2023-06-19       37      437760     46.9758    19.8994        214.9688   
65   2023-06-15       16      436492     45.7448    19.6851        215.1825   
128  2023-06-29       16      443556     48.4637    20.0436        213.9504   
122  2023-06-27       31      442600     48.2245    19.9494        214.1095   
102  2023-06-23       22      439452     46.9614    19.9469        214.6730   
27   2023-06-07        4      432030     46.4872    20.0829        216.0529   
28   2023-06-07       10      432032     46.4871    20.0828        216.0523   
62   2023-06-13       48      435837     47.6095    20.1163        215.2991   
66   2023-06-15       22      436495     45.7447    19.6851        215.1822   
73   2023-06-17       10      437196     46.6011    19.7767        215.0787   
94   2023-06-21       25      438700     47.1219    20.0134        214.8031   

     meanpred_train  resid_train  absresidmean_train  residmean_train  \
85         214.9800    1540.8985             19.9011           0.0035   
49         215.5993    3067.6114             20.2594           0.0071   
76         215.0559    3856.7440             19.9037           0.0088   
39         215.9079    1552.7407             20.2137           0.0036   
43         215.8760    1665.3969             20.1649           0.0038   
50         215.5714    3286.3075             20.1810           0.0076   
114        214.2532    2004.2458             19.7431           0.0045   
16         216.5204    2845.7522             20.2703           0.0066   
96         214.7833    2372.0742             19.8626           0.0054   
42         215.8817    2580.5889     

In [358]:
offersp_scores.pivot_table(index=['test_day'], aggfunc={'rmse_test':['mean','max','min'], 'size_test':'sum'})

rmse_test                     size_test
                 max       mean      min       sum
test_day                                          
2023-06-01   47.2631  14.323467   5.4973        64
2023-06-03   28.2386  17.682122   7.2144        67
2023-06-05   30.9652  13.159556   4.7213        88
2023-06-07   32.6543  10.783486   3.1262        37
2023-06-09   27.3505  14.229233   6.9261        60
2023-06-11   62.8443  18.456322   7.3257       309
2023-06-13  301.9327  56.207400  12.8735        65
2023-06-15   16.7519   9.954762   1.9838        57
2023-06-17   26.0980  12.851788   6.3116        27
2023-06-19   21.5872  11.444833   2.3149        39
2023-06-21   21.8867  12.987213   6.3996        89
2023-06-23   32.8186  15.434033   0.0552        64
2023-06-25  186.9261  34.130856   7.2307       366
2023-06-27   24.0554   9.725200   0.8297        50
2023-06-29   52.8346  17.526678   1.1732        27

In [359]:
offersp_scores.pivot_table(index=['test_sp'], aggfunc={'rmse_test':['mean','max','min'], 'size_test':'sum'})

rmse_test                    size_test
              max       mean     min       sum
test_sp                                       
4         62.8443  18.958371  3.2088       112
10        24.0554  14.223531  5.4973       102
16       186.9261  24.136393  1.1732       134
22        26.0980  13.714900  0.0552       205
25       301.9327  34.631129  3.9335       198
31        32.6543  15.246680  0.8297       190
37        17.0809  10.279847  2.3149       206
40        23.6055   9.878740  3.8374       150
48        47.2631  20.443714  3.1262       112

### - XGBOOST (OFFER MODEL - SP PAVG. PRICE MODEL)

In [360]:
file_paths = ['/content/drive/MyDrive/QEnergy/Results/XGBoost/XGBoost OFFER SP PAVG. PRICE MODEL OPT. - SP (scores MAPE).csv',
              '/content/drive/MyDrive/QEnergy/Results/XGBoost/Archive/Old/XGBoost OFFER SP PAVG. PRICE MODEL - SP (scores)+.csv',
              '/content/drive/MyDrive/QEnergy/Results/XGBoost/Archive/Old/XGBoost OFFER SP PAVG. PRICE MODEL OPT. - SP (scores)+.csv',
              '/content/drive/MyDrive/QEnergy/Results/XGBoost/Archive/Old/XGBoost OFFER SP PAVG. PRICE MODEL - SP (scores)+2.csv']

In [361]:

print('XGBOOST (OFFER MODEL - SP PAVG. PRICE MODEL)')

for x in file_paths:

  offerpavg_scores = pd.read_csv(x)

  offerpavg_rsme = offerpavg_scores['rmse_test'].mean()
  offerpavg_mae = offerpavg_scores['mae_test'].mean()
  offerpavg_meanreal = offerpavg_scores['meanreal_test'].mean()

  offerpavg_mape = (offerpavg_scores['mae_test']/offerpavg_scores['meanreal_test']).mean() * 100
  #offerpavg_mape = (offerpavg_mae/offerpavg_meanreal) * 100

  print('_'*30)
  print(f'Avg. RSME: {offerpavg_rsme:0.2f}')
  print(f'Avg. MAE: {offerpavg_mae:0.2f}')
  print(f'Avg. REAL BID PRICE: {offerpavg_meanreal:0.2f}')
  print(f'Avg. MAPE: {offerpavg_mape:0.2f} %')


XGBOOST (OFFER MODEL - SP PAVG. PRICE MODEL)
______________________________
Avg. RSME: 14.80
Avg. MAE: 11.89
Avg. REAL BID PRICE: 128.45
Avg. MAPE: 9.37 %
______________________________
Avg. RSME: 12.90
Avg. MAE: 10.63
Avg. REAL BID PRICE: 128.45
Avg. MAPE: 8.31 %
______________________________
Avg. RSME: 16.27
Avg. MAE: 12.94
Avg. REAL BID PRICE: 128.45
Avg. MAPE: 10.48 %
______________________________
Avg. RSME: 13.61
Avg. MAE: 10.82
Avg. REAL BID PRICE: 128.45
Avg. MAPE: 8.60 %


In [362]:
file_path = '/content/drive/MyDrive/QEnergy/Results/XGBoost/XGBoost OFFER SP PAVG. PRICE MODEL OPT. - SP (feat.importances).csv'
offerpavg_imp = pd.read_csv(file_path)

In [363]:
offerpavg_imp

feature   imp_sum  imp_2023-06-01  imp_2023-06-03  \
0                 sp3_6_pavgprice  0.219424        0.220154        0.221031   
1                  settPeriod_COS  0.154967        0.159906        0.171425   
2                       Day_ordin  0.066664        0.067780        0.062328   
3                      Coal_price  0.051583        0.042114        0.044233   
4                 LOC Center LONG  0.045762        0.046205        0.048129   
5                sp3_6_meanvolume  0.032955        0.033400        0.033998   
6                        DAMargin  0.028837        0.027459        0.031771   
7         WindOffshore_Generation  0.022327        0.021325        0.025011   
8            settlementDate_ordin  0.018843        0.017828        0.019334   
9                  LOC Center LAT  0.018816        0.027241        0.014381   
10               marketIndexPrice  0.018754        0.020617        0.018201   
11                    DAImbalance  0.018477        0.017676        0.019885   
12                    DATF_Demand  0.017510        0.017903        0.016143   
13                    Month_ordin  0.016491        0.014007        0.015158   
14                   NatGas_price  0.015634        0.014722        0.013959   
15              sp3_6_totalvolume  0.014762        0.015861        0.015078   
16                    Weekday_COS  0.014467        0.012726        0.014333   
17                        Weekend  0.013984        0.002574        0.004109   
18              marketIndexVolume  0.013864        0.013278        0.015907   
19              sp6_12_meanvolume  0.013568        0.008978        0.009999   
20                 BrentOil_price  0.013386        0.025102        0.011703   
21       BMU GSP Group Name_label  0.013292        0.012535        0.011882   
22                 settPeriod_SIN  0.012633        0.013044        0.011150   
23               Solar_Generation  0.012217        0.013528        0.013206   
24                        Day_COS  0.012054        0.012454        0.011130   
25                        Day_SIN  0.011885        0.013187        0.011444   
26  GenerationForecast_Production  0.011830        0.008096        0.008997   
27                 DAID_spnDemand  0.011390        0.008273        0.022041   
28               sp6_12_pavgprice  0.010808        0.010505        0.011646   
29         WindOnshore_Generation  0.010350        0.012962        0.005546   
30                      Month_SIN  0.010293        0.013817        0.011563   
31                    DANF_Demand  0.009383        0.011690        0.014437   
32             sp6_12_totalvolume  0.009083        0.010397        0.010175   
33             DAIG_spnGeneration  0.007777        0.006796        0.006931   
34                    Weekday_SIN  0.007762        0.008205        0.007311   
35                      Month_COS  0.006007        0.007197        0.003499   
36         settlementPeriod_ordin  0.005334        0.005232        0.004379   
37    SystemTotalLoad_Consumption  0.005020        0.004019        0.007589   
38                        Holiday  0.001808        0.001208        0.000961   

    imp_2023-06-05  imp_2023-06-07  imp_2023-06-09  imp_2023-06-11  \
0         0.213738        0.210362        0.222195        0.223468   
1         0.165679        0.146860        0.160823        0.156499   
2         0.056367        0.066841        0.068460        0.076781   
3         0.041889        0.051500        0.041219        0.042874   
4         0.047326        0.043097        0.045218        0.047239   
5         0.031577        0.034328        0.034082        0.030729   
6         0.029280        0.027921        0.031070        0.028827   
7         0.020552        0.018010        0.018915        0.026476   
8         0.019173        0.018374        0.021360        0.023657   
9         0.019553        0.014079        0.019549        0.015164   
10        0.016569        0.015620        0.019905        0.016934   
11        0.019948       

In [364]:
offerpavg_scores.sort_values(by=['rmse_test'], ascending=False).head(40)

test_day  test_sp  size_train  rmse_train  mae_train  meanreal_train  \
26   2023-06-05       48      225150     59.2881    22.7697        215.3958   
8    2023-06-01       48      223974     57.1201    22.8471        215.9052   
32   2023-06-07       31      225496     58.5821    22.9337        215.2351   
58   2023-06-13       25      227211     57.0079    22.6648        214.5926   
40   2023-06-09       25      225828     56.2184    22.4436        215.0850   
112  2023-06-25       25      229886     57.3743    22.6090        213.6780   
107  2023-06-23       48      229364     58.7921    22.6022        213.8579   
134  2023-06-29       48      231093     57.7585    22.5607        213.2685   
116  2023-06-25       48      230112     58.9206    22.6518        213.6017   
55   2023-06-13       10      227138     55.8128    22.6510        214.6197   
57   2023-06-13       22      227204     57.0087    22.6649        214.5961   
17   2023-06-03       48      224508     57.0958    22.7159        215.6720   
89   2023-06-19       48      228443     58.3477    22.5618        214.1731   
56   2023-06-13       16      227176     57.8786    22.8254        214.6070   
44   2023-06-09       48      225929     56.0679    22.5840        215.0376   
113  2023-06-25       31      229970     55.8564    22.4493        213.6496   
118  2023-06-27       10      230449     56.8636    22.3036        213.4853   
95   2023-06-21       31      228822     56.8665    22.4513        214.0428   
86   2023-06-19       31      228360     57.3562    22.6572        214.2026   
77   2023-06-17       31      228093     57.9906    22.5424        214.3052   
75   2023-06-17       22      228076     57.9927    22.5435        214.3127   
2    2023-06-01       16      223793     58.7094    22.9802        215.9893   
60   2023-06-13       37      227286     56.6006    22.7725        214.5637   
59   2023-06-13       31      227231     57.8058    22.7469        214.5847   
42   2023-06-09       37      225883     55.7551    22.5778        215.0574   
69   2023-06-15       37      227697     56.1738    22.6707        214.4261   
115  2023-06-25       40      230037     56.3822    22.4600        213.6250   
127  2023-06-29       10      230937     58.3718    22.5997        213.3209   
45   2023-06-11        4      226321     57.0925    22.6301        214.8819   
38   2023-06-09       16      225788     56.8557    22.8411        215.1049   
47   2023-06-11       16      226461     56.2563    22.5511        214.8323   
131  2023-06-29       31      231007     56.5901    22.4267        213.2966   
54   2023-06-13        4      227105     56.2060    22.7413        214.6320   
24   2023-06-05       37      225094     58.4612    22.7730        215.4203   
41   2023-06-09       31      225861     55.7576    22.5790        215.0674   
125  2023-06-27       48      230625     55.4326    22.3414        213.4242   
103  2023-06-23       25      229244     54.9688    22.3751        213.9016   
12   2023-06-03       22      224324     59.2939    22.8687        215.7568   
84   2023-06-19       22      228348     57.3577    22.6579        214.2074   
87   2023-06-19       37      228388     57.8810    22.7187        214.1917   

     meanpred_train  resid_train  absresidmean_train  residmean_train  \
26         215.3958    2474.3316             22.7697           0.0110   
8          215.9052    4795.0374             22.8471           0.0214   
32         215.2351    3607.2736             22.9337           0.0160   
58         214.5926    2812.2227             22.6648           0.0124   
40         215.0850    3280.6084             22.4436           0.0145   
112        213.6780    2667.1344             22.6090           0.0116   
107        213.8579    2944.6356             22.6022           0.0128   
134        213.2685    4145.4745             22.5607           0.0179   
116        213.6017    2146.4228             22.6518           0.0093   
55         214.6197    2088.6308     

In [365]:
offerpavg_scores.sort_values(by=['rmse_test'], ascending=False).tail(40)

test_day  test_sp  size_train  rmse_train  mae_train  meanreal_train  \
22   2023-06-05       25      225039     57.9293    22.7614        215.4442   
43   2023-06-09       40      225900     56.0713    22.5859        215.0498   
21   2023-06-05       22      225016     59.4604    22.8401        215.4545   
99   2023-06-23        4      229190     55.1997    22.2469        213.9219   
3    2023-06-01       22      223828     57.2467    22.8097        215.9732   
37   2023-06-09       10      225759     56.5590    22.6837        215.1171   
70   2023-06-15       40      227722     56.1709    22.6694        214.4195   
20   2023-06-05       16      224994     56.4785    22.7830        215.4641   
67   2023-06-15       25      227670     56.1769    22.6721        214.4348   
51   2023-06-11       37      226644     57.3904    22.6362        214.7622   
50   2023-06-11       31      226593     57.2446    22.6765        214.7816   
6    2023-06-01       37      223919     58.1914    22.8170        215.9311   
130  2023-06-29       25      230984     58.3663    22.5977        213.3052   
52   2023-06-11       40      226683     57.4844    22.6057        214.7488   
117  2023-06-27        4      230426     56.7800    22.5238        213.4930   
102  2023-06-23       22      229240     54.9692    22.3754        213.9031   
101  2023-06-23       16      229227     54.9708    22.3765        213.9075   
109  2023-06-25       10      229715     55.2863    22.2416        213.7382   
18   2023-06-05        4      224942     59.1585    22.8356        215.4866   
110  2023-06-25       16      229773     57.0119    22.5358        213.7171   
133  2023-06-29       40      231047     56.5518    22.4252        213.2830   
49   2023-06-11       25      226536     59.4261    22.6750        214.8043   
128  2023-06-29       16      230961     58.3691    22.5992        213.3140   
80   2023-06-17       48      228146     59.4808    22.7235        214.2863   
79   2023-06-17       40      228119     57.8202    22.5448        214.2951   
78   2023-06-17       37      228116     58.2683    22.5636        214.2962   
68   2023-06-15       31      227687     56.1748    22.6712        214.4295   
82   2023-06-19       10      228319     57.2850    22.6638        214.2196   
114  2023-06-25       37      230022     56.9015    22.4742        213.6302   
65   2023-06-15       16      227666     56.1773    22.6724        214.4355   
91   2023-06-21       10      228782     56.8712    22.4535        214.0568   
129  2023-06-29       22      230975     58.3675    22.5983        213.3087   
34   2023-06-07       40      225527     58.1007    22.8281        215.2206   
35   2023-06-07       48      225558     58.0967    22.8256        215.2061   
27   2023-06-07        4      225428     58.6855    22.9637        215.2649   
28   2023-06-07       10      225430     58.6853    22.9637        215.2638   
62   2023-06-13       48      227354     56.3554    22.6993        214.5414   
66   2023-06-15       22      227669     56.1770    22.6722        214.4350   
73   2023-06-17       10      228021     57.1726    22.5103        214.3312   
94   2023-06-21       25      228812     58.7890    22.5123        214.0472   

     meanpred_train  resid_train  absresidmean_train  residmean_train  \
22         215.4442    3350.8497             22.7614           0.0149   
43         215.0498    2416.0466             22.5859           0.0107   
21         215.4545    2836.9665             22.8401           0.0126   
99         213.9219    1312.7707             22.2469           0.0057   
3          215.9732    4118.6260             22.8097           0.0184   
37         215.1171    1572.4212             22.6837           0.0070   
70         214.4195    2725.4904             22.6694           0.0120   
20         215.4641    4868.6872             22.7830           0.0216   
67         214.4348    2727.3627             22.6721           0.0120   
51         214.7622    2266.8077     

In [366]:
offerpavg_scores.pivot_table(index=['test_day'], aggfunc={'rmse_test':['mean','max','min'], 'size_test':'sum'})

rmse_test                     size_test
                 max       mean      min       sum
test_day                                          
2023-06-01   57.0846  16.492322   5.0776        44
2023-06-03   28.6427  13.862211   8.2714        48
2023-06-05   60.7089  15.036933   3.6179        52
2023-06-07   38.7341  11.740286   0.3284        22
2023-06-09   34.3364  16.915489   5.7989        43
2023-06-11   19.6544   9.683678   3.1015        99
2023-06-13   35.3951  24.477187  12.9791        47
2023-06-15   20.5554   8.786212   1.5826        33
2023-06-17   23.0741   9.316187   2.5991        22
2023-06-19   26.9510  14.813722   2.2923        29
2023-06-21   23.3990  10.946513   1.3981        45
2023-06-23   33.4693  11.933500   4.2967        38
2023-06-25   33.8707  16.314056   2.0701        87
2023-06-27   23.5290  10.900900   4.3412        34
2023-06-29   33.2797  12.477733   1.0698        23

In [367]:
offerpavg_scores.pivot_table(index=['test_sp'], aggfunc={'rmse_test':['mean','max','min'], 'size_test':'sum'})

rmse_test                    size_test
              max       mean     min       sum
test_sp                                       
4         19.6544  11.723671  3.6179        70
10        29.6485  11.412808  1.3981        58
16        25.9887  11.211953  1.5826        74
22        29.0015  11.373636  1.0698        69
25        35.3951  14.146643  3.1015        71
31        38.7341  16.741820  2.4550        77
37        22.4141  12.458020  2.0701        96
40        20.4638   8.727880  0.8756        83
48        60.7089  24.956343  0.3284        68